hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -323,21 +315,67 @@ public synchronized void start() {
     }
 
     public void tryConnectToAllClusterMembers(boolean sync) {
-        if (!isSmartRoutingEnabled) {
-            return;
+        if (isSmartRoutingEnabled && sync) {
+            connectToAllClusterMembers();
         }
+    }
 
-        if (sync) {
-            for (Member member : client.getClientClusterService().getMemberList()) {
-                try {
-                    getOrConnectToMember(member, false);
-                } catch (Exception e) {
-                    EmptyStatement.ignore(e);
+    private void clusterThreadLoop() {
+        while (isAlive.get()) {
+            try {
+                Runnable runnable = taskQueue.poll(1, TimeUnit.SECONDS);
+                if (!isAlive.get()) {
+                    break;
+                }
+                if (runnable != null) {
+                    runnable.run();
+                } else if (isSmartRoutingEnabled) {
+                    connectToAllClusterMembers();
                 }
+            } catch (Exception e) {
+                logger.warning("Exception occurred while executing cluster task", e);
+            }
+        }
+    }
+
+    private void handleConnectionClose(TcpClientConnection connection) {
```

Comment:
So, with this mechanism, the connections will be closed, but they can potentially stay in the active connections map for some time, since this will be executed in the cluster executor. Isn't this a problem?

Reply:
We just increased the time that it can stay in active connections. It was always possible to have a closed connections in active connections. In theory, it should be safe.